<a href="https://colab.research.google.com/github/saifinuha/data-science-2026/blob/main/Pertemuan2_Saifin_Nuha_240401010257.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [10]:
# Membuat list
data = [10, 20, 'Python', True, 3.14]
# Akses & slicing
print(data[0]) # → 10
print(data[-1]) # → 3.14 (dari belakang)
print(data[1:3]) # → [20, 'Python']
# Manipulasi
data.append(99) # tambah di akhir
data.insert(1, 55) # sisipkan di posisi 1
data.remove(20) # hapus nilai pertama = 20
# data.sort() # urutkan (jika tipe sama) - Karena tipe tidak sama, tidak jalan

# Tuple
mahasiswa = {
'nama': 'Budi Santoso',
'nim': '2024001',
'prodi': 'Informatika',
'ipk': 3.85
}
# Akses nilai
print(mahasiswa['nama']) # → 'Budi Santoso'
print(mahasiswa.get('usia', 0)) # default jika key tak ada
# Update dan hapus
mahasiswa['angkatan'] = 2024 # tambah key baru
del mahasiswa['nim'] # hapus key

import numpy as np
# Dari list Python
a1D = np.array([1, 2, 3, 4, 5]) # 1D
a2D = np.array([[1,2,3],[4,5,6]]) # 2D
# Array dengan nilai inisial
zeros = np.zeros((3, 4)) # nol semua
ones = np.ones((2, 3)) # satu semua
eye = np.eye(4) # matriks identitas
full = np.full((2, 2), 7) # nilai konstan
# Sequence
arange = np.arange(0, 20, 2) # 0,2,4,...18
linsp = np.linspace(0, 1, 11) # 11 titik merata
# Array acak (reproducible)
rng = np.random.default_rng(seed=42)
rand_int = rng.integers(0, 100, (3, 4))
rand_flt = rng.normal(loc=0, scale=1, size=(100,))

# Operasi aritmetika vektorisasi
a = np.array([1, 2, 3, 4, 5])
b = np.array([10, 20, 30, 40, 50])
print(a + b) # → [11 22 33 44 55]
print(a * b) # → [10 40 90 160 250]
print(a ** 2) # → [1 4 9 16 25]
# Universal functions (ufuncs)
np.sqrt(a) # akar kuadrat tiap elemen
np.exp(a) # eksponensial (e^x)
np.log(b) # logaritma natural
# Fungsi statistik
data = rng.normal(0, 1, 1000)
print(f'Mean: {np.mean(data):.4f}')
print(f'Std: {np.std(data):.4f}')
print(f'Min: {np.min(data):.4f}')
print(f'Max: {np.max(data):.4f}')
print(f'Q1: {np.percentile(data,25):.4f}')
print(f'Q3: {np.percentile(data,75):.4f}')

import pandas as pd

# Series dari list — index default 0,1,2...
s1 = pd.Series([85, 92, 78, 95, 88])
# Series dengan index kustom
nilai = pd.Series(
[85, 92, 78, 95],
index=['Alice','Bob','Caca','Dina']
)
print(nilai['Bob']) # → 92
print(nilai[nilai>80]) # filter
print(nilai.mean()) # → 87.5
print(nilai.describe()) # statistik lengkap

# Membuat DataFrame dari dictionary
data = {
'nama': ['Ali','Budi','Caca','Dina'],
'usia': [22, 25, 21, 23],
'kota': ['Jakarta','Bandung','Surabaya','Medan'],
'ipk': [3.8, 3.5, 3.9, 3.7]
}
df = pd.DataFrame(data)
# Informasi dasar
print(df.shape) # → (4, 4)
print(df.dtypes) # tipe data per kolom
print(df.columns) # nama-nama kolom

# Memuat dari file lokal
# df = pd.read_csv('data/penjualan.csv')
# Memuat dari URL (dataset publik)
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv" # Changed to raw URL
df = pd.read_csv(url)
# Parameter penting
# df = pd.read_csv('data.csv',
# sep=',', # pemisah kolom (default ',')
# encoding='utf-8', # encoding file
# index_col='PassengerId', # kolom sebagai index
# na_values=['N/A','none',''] # nilai kosong
# )
# Eksplorasi standar setelah load
df.head() # 5 baris pertama
df.info() # tipe data & missing
df.describe() # statistik deskriptif

# Contoh: Dataset Titanic - sudah diload di atas
# df = pd.read_csv(titanic_url)
# Filter satu kondisi
penumpang_selamat = df[df['Survived'] == 1]
penumpang_wanita = df[df['Sex'] == 'female']
# Filter kondisi ganda (& = AND, | = OR)
wanita_kelas1 = df[(df['Sex']=='female') &
(df['Pclass']==1)]
# Seleksi kolom tertentu
df[['Name','Age','Sex','Survived']]

# Pola dasar: groupby → aggregation
df.groupby('Pclass')['Survived'].mean()
# Agregasi multiple kolom & fungsi
hasil = df.groupby('Pclass').agg(
total = ('PassengerId', 'count'),
selamat = ('Survived', 'sum'),
pct_selamat= ('Survived', 'mean'),
rata_usia = ('Age', 'mean'),
max_fare = ('Fare', 'max')
)
hasil = hasil.round(3)
print(hasil)
# GroupBy multi-kolom
df.groupby(['Pclass','Sex'])['Survived']\
.mean().unstack().round(3)

10
3.14
[20, 'Python']
Budi Santoso
0
[11 22 33 44 55]
[ 10  40  90 160 250]
[ 1  4  9 16 25]
Mean: -0.0163
Std: 1.0008
Min: -3.6484
Max: 3.1789
Q1: -0.6897
Q3: 0.6163
92
Alice    85
Bob      92
Dina     95
dtype: int64
87.5
count     4.000000
mean     87.500000
std       7.593857
min      78.000000
25%      83.250000
50%      88.500000
75%      92.750000
max      95.000000
dtype: float64
(4, 4)
nama     object
usia      int64
kota     object
ipk     float64
dtype: object
Index(['nama', 'usia', 'kota', 'ipk'], dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-n

Sex,female,male
Pclass,,
1,0.968,0.369
2,0.921,0.157
3,0.500,0.135


Q1 Berapa jumlah baris dan kolom pada dataset Titanic? Kolom apa saja yang tersedia?

Q2 Kolom mana saja yang memiliki missing values? Berapa jumlah dan proporsinya (%) per kolom?

Q3 Berapa persen penumpang yang selamat (Survived=1)? Gunakan operasi NumPy
atau Pandas.

Q4 Filter penumpang wanita dari kelas 1 (Pclass=1). Berapa jumlahnya dan berapa rata-rata usianya?

Q5 Gunakan groupby untuk menghitung tingkat keselamatan (mean Survived) per kelas penumpang. Kelas mana yang paling tinggi?

Q6 Buat analisis tambahan berdasarkan variabel lain yang Anda anggap menarik.
Sertakan interpretasi singkat.

In [11]:
# Hitung rata-rata usia untuk wanita kelas 1
rata_rata_usia_wanita_kelas1 = wanita_kelas1['Age'].mean()
print(f"Rata-rata usia penumpang wanita kelas 1: {rata_rata_usia_wanita_kelas1:.2f} tahun")

Rata-rata usia penumpang wanita kelas 1: 34.61 tahun


### Jawaban:

**Q1: Berapa jumlah baris dan kolom pada dataset Titanic? Kolom apa saja yang tersedia?**
*   **Jumlah Baris:** 891
*   **Jumlah Kolom:** 12
*   **Kolom yang Tersedia:** `PassengerId`, `Survived`, `Pclass`, `Name`, `Sex`, `Age`, `SibSp`, `Parch`, `Ticket`, `Fare`, `Cabin`, `Embarked`.

**Q2: Kolom mana saja yang memiliki missing values? Berapa jumlah dan proporsinya (%) per kolom?**
Dari `df.info()`:
*   **Age:** 714 non-null dari 891 baris. Missing: `891 - 714 = 177` (`177 / 891 * 100 = 19.87%`).
*   **Cabin:** 204 non-null dari 891 baris. Missing: `891 - 204 = 687` (`687 / 891 * 100 = 77.10%`).
*   **Embarked:** 889 non-null dari 891 baris. Missing: `891 - 889 = 2` (`2 / 891 * 100 = 0.22%`).

**Q3: Berapa persen penumpang yang selamat (Survived=1)?**
*   Dari `df.describe()` atau `df.groupby('Pclass')['Survived'].mean()` secara keseluruhan, rata-rata `Survived` adalah `0.383838`. Ini berarti sekitar **38.38%** penumpang selamat.

**Q4: Filter penumpang wanita dari kelas 1 (Pclass=1). Berapa jumlahnya dan berapa rata-rata usianya?**
*   **Jumlah:** Terdapat 94 penumpang wanita dari kelas 1 (dari variabel `wanita_kelas1`).
*   **Rata-rata Usia:** (akan dihitung dari sel kode di atas)

**Q5: Gunakan groupby untuk menghitung tingkat keselamatan (mean Survived) per kelas penumpang. Kelas mana yang paling tinggi?**
Dari hasil `df.groupby('Pclass')['Survived'].mean()`:
```
        total  selamat  pct_selamat  rata_usia  max_fare
Pclass                                                  
1         216      136        0.630     38.233   512.329
2         184       87        0.473     29.878    73.500
3         491      119        0.242     25.141    69.550
```
*   Tingkat keselamatan tertinggi ada pada **Kelas 1** dengan rata-rata 0.630 (63.0%).

**Q6: Buat analisis tambahan berdasarkan variabel lain yang Anda anggap menarik. Sertakan interpretasi singkat.**
Analisis tambahan dapat dilakukan dengan melihat tingkat keselamatan berdasarkan kombinasi Kelas Penumpang (`Pclass`) dan Jenis Kelamin (`Sex`).
Dari hasil `df.groupby(['Pclass','Sex'])['Survived'].mean().unstack().round(3)`:
```
Sex     female   male
Pclass              
1        0.968  0.369
2        0.921  0.157
3        0.500  0.135
```
*   **Interpretasi Singkat:** Terlihat bahwa tingkat keselamatan wanita jauh lebih tinggi dibandingkan pria di semua kelas. Wanita di kelas 1 memiliki tingkat keselamatan tertinggi (96.8%), sementara pria di kelas 3 memiliki tingkat keselamatan terendah (13.5%). Ini menunjukkan bahwa jenis kelamin dan kelas sosial merupakan faktor yang sangat signifikan dalam tingkat kelangsungan hidup pada bencana Titanic.